jmxtrans-core/src/main/java/com/googlecode/jmxtrans/connections/JMXConnection.java

```diff
@@ -30,31 +31,81 @@
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import javax.annotation.concurrent.ThreadSafe;
+import javax.management.AttributeChangeNotification;
+import javax.management.InstanceNotFoundException;
 import javax.management.MBeanServerConnection;
+import javax.management.Notification;
+import javax.management.NotificationFilter;
+import javax.management.ObjectName;
 import javax.management.remote.JMXConnector;
 import java.io.Closeable;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
 
 @ToString
 @ThreadSafe
 public class JMXConnection implements Closeable {
-	@Nullable private final JMXConnector connector;
-	@Nonnull @Getter private final MBeanServerConnection mBeanServerConnection;
+	@Nullable
+	private final JMXConnector connector;
+	@Nonnull
+	@Getter
+	private final MBeanServerConnection mBeanServerConnection;
 	private boolean markedAsDestroyed;
 	private static final Logger logger = LoggerFactory.getLogger(JMXConnection.class);
 
+	private final IdentityHashMap<Object, QueryNotificationListener> notificationListeners = new IdentityHashMap<>();
+
 	public JMXConnection(@Nullable JMXConnector connector, @Nonnull MBeanServerConnection mBeanServerConnection) {
 		this.connector = connector;
 		this.mBeanServerConnection = mBeanServerConnection;
 		this.markedAsDestroyed = false;
+	}
+
+	private static NotificationFilter getNotificationFilter(final ImmutableList<String> attributes) {
+		return new NotificationFilter() {
+			@Override
+			public boolean isNotificationEnabled(Notification notification) {
+				if (notification instanceof AttributeChangeNotification) {
+					// Check if subscribed attribute
+					AttributeChangeNotification changeNotification = (AttributeChangeNotification) notification;
+					return attributes.contains(changeNotification.getAttributeName());
+				}
+				return false;
```

Comment:
Minor style issue: I prefer to have preconditions with early return instead of nested blocks. In this case, I think it would be nicer to have:

```
if (! notification instanceof AttributeChangeNotification) return false;
[...]
```

Reply:
Besides the style issue, do you think that limiting the support to AttributeChangeNotifications is OK for now?

Reply:
I have a feeling that supporting other kind of notification could have value. But starting with just AttributeChange and building up from there is perfectly fine. We can always improve later if there is a need.